internal/engine/experiment/torsf/torsf.go

```diff
@@ -101,47 +127,127 @@ func (m *Measurer) Run(
-func (m *Measurer) run(ctx context.Context,
-	sess model.ExperimentSession, testkeys *TestKeys, errch chan<- error) {
-	sfdialer := &ptx.SnowflakeDialer{}
+// setup prepares for running the torsf experiment. Returns a valid ptx listener
+// and snowflake dialer on success. Returns an error on failure. On success,
+// remember to Stop the ptx listener when you're done.
+func (m *Measurer) setup(logger model.Logger) (*ptx.Listener, *ptx.SnowflakeDialer, error) {
+	rm, err := ptx.NewSnowflakeRendezvousMethod(m.config.RendezvousMethod)
+	if err != nil {
+		// cannot run the experiment with unknown rendezvous method
+		return nil, nil, err
+	}
+	sfdialer := ptx.NewSnowflakeDialerWithRendezvousMethod(rm)
 	ptl := &ptx.Listener{
 		PTDialer: sfdialer,
-		Logger:   sess.Logger(),
+		Logger:   logger,
 	}
 	if err := m.startListener(ptl.Start); err != nil {
-		testkeys.Failure = archival.NewFailure(err)
 		// This error condition mostly means "I could not open a local
 		// listening port", which strikes as fundamental failure.
-		errch <- err
-		return
+		return nil, nil, err
 	}
-	defer ptl.Stop()
+	logger.Infof("torsf: rendezvous method: '%s'", m.config.RendezvousMethod)
+	return ptl, sfdialer, nil
+}
+
+// bootstrap runs the bootstrap.
+func (m *Measurer) bootstrap(ctx context.Context, sess model.ExperimentSession,
+	out chan<- *TestKeys, ptl *ptx.Listener, sfdialer *ptx.SnowflakeDialer) {
+	tk := &TestKeys{
+		BootstrapTime:     0,
+		Failure:           nil,
+		PersistentDatadir: !m.config.DisablePersistentDatadir,
+		RendezvousMethod:  sfdialer.RendezvousMethod.Name(),
+	}
+	sess.Logger().Infof(
+		"torsf: disable persistent datadir: %+v", m.config.DisablePersistentDatadir)
+	defer func() {
+		out <- tk
+	}()
 	tun, err := m.startTunnel()(ctx, &tunnel.Config{
 		Name:      "tor",
 		Session:   sess,
-		TunnelDir: path.Join(sess.TempDir(), "torsf"),
+		TunnelDir: path.Join(m.baseTunnelDir(sess), "torsf"),
 		Logger:    sess.Logger(),
 		TorArgs: []string{
 			"UseBridges", "1",
 			"ClientTransportPlugin", ptl.AsClientTransportPluginArgument(),
 			"Bridge", sfdialer.AsBridgeArgument(),
 		},
 	})
+	m.readTorLogs(sess.Logger(), tk, tun)
 	if err != nil {
 		// Note: archival.NewFailure scrubs IP addresses
-		testkeys.Failure = archival.NewFailure(err)
-		// This error condition means we could not bootstrap with snowflake
-		// for $reasons, so the experiment didn't fail, rather it did record
-		// that something prevented snowflake from running.
-		errch <- nil
+		tk.Failure = archival.NewFailure(err)
 		return
 	}
 	defer tun.Stop()
-	testkeys.BootstrapTime = tun.BootstrapTime().Seconds()
-	errch <- nil
+	tk.BootstrapTime = tun.BootstrapTime().Seconds()
+}
+
+// readTorLogs attempts to read and include the tor logs into
+// the test keys if this operation is possible.
+//
+// This function aims to _only_ include:
+//
+// 1. information on the tor version by intercepting the line that
+// writes which tor version is opening a log file;
```

Comment:
I tried running this test with:
```
./miniooni -ODisablePersistentDatadir=true \
           -ORendezvousMethod=amp \
           -ODisableProgress=true torsf
```

And didn't find in the `tor_logs` list the line referring to the specific tor version that was being used.

I wonder if we should rather obtain the tor_version from the control port and embed it as separate key in the data format (see: https://github.com/cretz/bine/blob/f9f678b84ccaed1c3f919406b44a7d95ddbd6d47/control/cmd_protocolinfo.go#L13).

Reply:
Good suggestion, thank you! I have implemented that!